omniduct/databases/base.py

```diff
@@ -258,22 +236,44 @@ def render_template(self, name_or_statement, context=None, by_name=False):
         else:
             statement = name_or_statement
 
-        if not context:
-            context = {}
-
-        # Substitute in any other named statements recursively
-        while '{{{' in statement or '{{%' in statement:
-            statement = Template(statement,
-                                 block_start_string='{{%',
-                                 block_end_string='%}}',
-                                 variable_start_string='{{{',
-                                 variable_end_string='}}}',
-                                 comment_start_string='{{#',
-                                 comment_end_string='#}}',
-                                 undefined=StrictUndefined).render(getattr(self, '_templates', {}))
-
-        # Evaluate final template in provided context
-        statement = Template(statement, undefined=StrictUndefined).render(context)
+        try:
+            from sqlalchemy.sql.base import Executable
+            if isinstance(statement, Executable):
+                statement = str(statement.compile(compile_kwargs={"literal_binds": True}))
+        except ImportError:
+            pass
+
+        # If templating enabled, render template
+        if template:
+
+            if context is None:
+                context = {}
+
+            template_context = {}
+            template_context.update(self._template_context)  # default context
+            template_context.update(context)
+            intersection = set(self._template_context.keys()) & set(context.keys())
+            if intersection:
+                logger.warning(
+                    "The following default template context keys have been overridden "
+                    "by the local context: {}."
+                    .format(intersection)
+                )
+
+            # Substitute in any other named statements recursively
+            while '{{{' in statement or '{{%' in statement:
+                statement = Template(statement,
+                                     block_start_string='{{%',
+                                     block_end_string='%}}',
+                                     variable_start_string='{{{',
+                                     variable_end_string='}}}',
+                                     comment_start_string='{{#',
+                                     comment_end_string='#}}',
+                                     undefined=StrictUndefined).render(getattr(self, '_templates', {}))
+
+            while '{{' in statement or '{%' in statement:
```

Comment:
FYI this is new as well but allows for variables to reference other elements in the context

Reply:
Hmmm... Can you give me an example of where this might be useful? The only things I can think of are better served by using local variables in python or nested templates.

Reply:
Since this is somewhat orthogonal to this change - I'm going to remove this and we can discuss if it comes up elsewhere